src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,770 @@
+                    credentials::uri(""),
+                    ter(temMALFORMED));
+            }
+
+            {
+                testcase("Credentials fail, expiration in the past.");
+                auto jv = credentials::create(subject, issuer, credType);
+                // current time in ripple epoch - 1s
+                uint32_t const t = env.now().time_since_epoch().count() - 1;
+                jv[sfExpiration.jsonName] = t;
+                env(jv, ter(tecEXPIRED));
+            }
+
+            {
+                testcase("Credentials fail, invalid fee.");
+
+                auto jv = credentials::create(subject, issuer, credType);
+                jv[jss::Fee] = -1;
+                env(jv, ter(temBAD_FEE));
+            }
+
+            {
+                testcase("Credentials fail, duplicate.");
+                auto const jv = credentials::create(subject, issuer, credType);
+                env(jv);
+                env.close();
+                env(jv, ter(tecDUPLICATE));
+                env.close();
+            }
+        }
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            env.fund(XRP(5000), issuer);
+            env.close();
+
+            {
+                testcase("Credentials fail, subject doesn't exist.");
+                auto const jv = credentials::create(subject, issuer, credType);
+                env(jv, ter(tecNO_TARGET));
+            }
+        }
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            auto const reserve = drops(env.current()->fees().accountReserve(0));
+            env.fund(reserve, subject, issuer);
+            env.close();
+
+            testcase("Credentials fail, not enough reserve.");
+            {
+                auto const jv = credentials::create(subject, issuer, credType);
+                env(jv, ter(tecINSUFFICIENT_RESERVE));
+                env.close();
+            }
+        }
+    }
+
+    void
+    testAcceptFailed(FeatureBitset features)
+    {
+        using namespace jtx;
+
+        const char credType[] = "abcde";
+        Account const issuer{"issuer"};
+        Account const subject{"subject"};
+        Account const other{"other"};
+
+        {
+            Env env{*this, features};
+
+            env.fund(XRP(5000), subject, issuer);
+
+            {
+                testcase("CredentialsAccept fail, Credential doesn't exist.");
+                env(credentials::accept(subject, issuer, credType),
+                    ter(tecNO_ENTRY));
+                env.close();
+            }
+
+            {
+                testcase("CredentialsAccept fail, invalid Issuer account.");
+                auto jv = credentials::accept(subject, issuer, credType);
+                jv[jss::Issuer] = to_string(xrpAccount());
+                env(jv, ter(temINVALID_ACCOUNT_ID));
+                env.close();
+            }
+        }
```

Comment:
Why not have this block in the same fully-funded block below (line 549), so that you don't need to spin up an extra `Env` (a bit more performant)?

Reply:
To not interfere with already created / accepted credentials